src/NHibernate/AdoNet/ConnectionManager.cs

```diff
@@ -312,117 +300,151 @@ public void GetObjectData(SerializationInfo info, StreamingContext context)
 
 		void IDeserializationCallback.OnDeserialization(object sender)
 		{
-			batcher = Factory.Settings.BatcherFactory.CreateBatcher(this, interceptor);
+			_batcher = Factory.Settings.BatcherFactory.CreateBatcher(this, _interceptor);
 		}
 
 		#endregion
 
 		public ITransaction BeginTransaction(IsolationLevel isolationLevel)
 		{
 			Transaction.Begin(isolationLevel);
-			return transaction;
+			return _transaction;
 		}
 
 		public ITransaction BeginTransaction()
 		{
 			Transaction.Begin();
-			return transaction;
+			return _transaction;
 		}
 
 		public ITransaction Transaction
 		{
 			get
 			{
-				if (transaction == null)
+				if (_transaction == null)
 				{
-					transaction = Factory.TransactionFactory.CreateTransaction(session);
+					_transaction = Factory.TransactionFactory.CreateTransaction(_session);
 				}
-				return transaction;
+				return _transaction;
 			}
 		}
 
 		public void AfterNonTransactionalQuery(bool success)
 		{
-			log.Debug("after autocommit");
+			_log.Debug("After autocommit");
 		}
 
 		private bool IsAfterTransactionRelease
-		{
-			get { return connectionReleaseMode == ConnectionReleaseMode.AfterTransaction; }
-		}
+			=> _connectionReleaseMode == ConnectionReleaseMode.AfterTransaction;
 
 		private bool IsOnCloseRelease
-		{
-			get { return connectionReleaseMode == ConnectionReleaseMode.OnClose; }
-		}
+			=> _connectionReleaseMode == ConnectionReleaseMode.OnClose;
 
 		private bool IsAggressiveRelease
-		{
-			get
-			{
-				if (connectionReleaseMode == ConnectionReleaseMode.AfterTransaction)
-				{
-					return !IsInActiveTransaction;
-				}
-				return false;
-			}
-		}
+			=> _connectionReleaseMode == ConnectionReleaseMode.AfterTransaction && !IsInActiveTransaction;
 
 		public ISessionFactoryImplementor Factory
-		{
-			get { return session.Factory; }
-		}
+			=> _session.Factory;
 
 		public bool IsReadyForSerialization
 		{
 			get
 			{
-				if (ownConnection)
-				{
-					return connection == null && !batcher.HasOpenResources;
-				}
-				else
+				if (_ownConnection)
 				{
-					return connection == null;
+					return _connection == null && !_batcher.HasOpenResources;
 				}
+				return _connection == null;
 			}
 		}
 
 		/// <summary> The batcher managed by this ConnectionManager. </summary>
 		public IBatcher Batcher
+			=> _batcher;
+
+		public DbCommand CreateCommand()
 		{
-			get { return batcher; }
+			var result = GetConnection().CreateCommand();
+			Transaction.Enlist(result);
+			return result;
 		}
 
+		/// <summary>
+		/// Enlist the connection into provided transaction if the connection should be enlisted.
+		/// Does nothing in case an explicit transaction ongoing.
+		/// </summary>
+		/// <param name="transaction">The transaction in which the connection should be enlisted.</param>
+		public void EnlistIfRequired(System.Transactions.Transaction transaction)
```

Comment:
Previously NHibernate was relying solely on connection auto-enlistment. This could not work with `OnClose` connection release mode, and this was sometimes failing due to some race condition causing the connection to have been un-enlisted soon enough from the transaction end. (In addition to not have been released and re-acquired.)
Enlisting explicitly avoids the "late un-enlistment" trouble.